gapis/api/transform2/capture_log.go

```diff
@@ -0,0 +1,130 @@
+
+import (
+	"context"
+	"os"
+
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/core/memory/arena"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/capture"
+)
+
+type captureLog struct {
+	file   *os.File
+	header *capture.Header
+	cmds   []api.Cmd
+	ids    map[api.CmdID]api.CmdID
+}
+
+// NewCaptureLog returns a Transformer that will log all commands passed through it
+// to the capture file at path.
+func NewCaptureLog(ctx context.Context, sourceCapture *capture.GraphicsCapture, path string) *captureLog {
+	f, err := os.Create(path)
+	if err != nil {
+		log.E(ctx, "Failed to create replay capture file %v: %v", path, err)
+		return nil
+	}
+	return &captureLog{
+		file:   f,
+		header: sourceCapture.Header,
+		cmds:   []api.Cmd{},
+		ids:    map[api.CmdID]api.CmdID{},
+	}
+}
+
+func (logTransform *captureLog) RequiresAccurateState() bool {
+	return false
+}
+
+func (logTransform *captureLog) BeginTransform(ctx context.Context, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	return inputCommands, nil
+}
+
+func (logTransform *captureLog) ClearTransformResources(ctx context.Context) {
+	// Do nothing
+}
+
+func (logTransform *captureLog) TransformCommand(ctx context.Context, id api.CmdID, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	// Melih TODO: We cannot have this check anymore.
+	// We can make an approximation that first command in the array likely to be a real command
+	// but it's still may not be true
+
+	// Don't write out the custom commands (e.g. replay.Custom)
+	// if cmd.API() != nil {
+	// 	if id.IsReal() {
+	// 		t.ids[id] = api.CmdID(len(t.cmds))
+	// 	}
+	// 	t.cmds = append(t.cmds, cmd)
+	// }
+
+	if len(inputCommands) > 0 {
+		logTransform.ids[id] = api.CmdID(len(logTransform.cmds))
+		logTransform.cmds = append(logTransform.cmds, inputCommands[0])
+	}
+
+	return inputCommands, nil
+}
+
+func (logTransform *captureLog) EndTransform(ctx context.Context, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	a := arena.New()
+	defer a.Dispose()
+
+	for idx := range logTransform.cmds {
+		cmd := logTransform.cmds[idx].Clone(a)
+		logTransform.cmds[idx] = cmd
+
+		// Melih TODO: We cannot check this anymore
+		// Due to this issue, probably functionality is broken
+
+		// if caller := cmd.Caller(); caller.IsReal() {
+		// 	if id, ok := t.ids[caller]; ok {
+		// 		cmd.SetCaller(id)
+		// 	} else {
+		// 		log.W(ctx, "Callee without caller %v: %v", caller, cmd)
+		// 		cmd.SetCaller(api.CmdNoID)
+		// 	}
+		// }
+
+		caller := cmd.Caller()
+		if id, ok := logTransform.ids[caller]; ok {
+			cmd.SetCaller(id)
+		} else {
+			log.W(ctx, "Callee without caller %v: %v", caller, cmd)
+
+			// Melih TODO: We are planning to remove api.CmdNoID
+			// I have decided put 0 here but it's just any arbitrary number.
+			// Object if you think this is wrong.
+
+			// cmd.SetCaller(api.CmdNoID)
+			cmd.SetCaller(0)
```

Comment:
Yes, I think this is wrong. We were trying to get away from this "magic number command id" stuff